zict/tests/test_func.py

```diff
@@ -46,3 +52,40 @@ def test_mapping():
     z = Func(rotl, rotr, d)
     utils_test.check_mapping(z)
     utils_test.check_closing(z)
+
+
+@pytest.mark.parametrize("wrapped_cls", [MutableMapping, ZictBase])
+def test_update_descopes_early(wrapped_cls):
+    """Test that Func.update() descopes the output of self.dump as soon as it can, if
+    the wrapped mapping allows, and doesn't store everything into a list.
+    """
+
+    class Dumped:
+        n = 0
+
+        def __init__(self):
+            gc.collect()  # Only necessary on pypy
+            Dumped.n += 1
+            assert Dumped.n < 3
+
+        def __del__(self):
+            Dumped.n -= 1
+
+    class Dummy(wrapped_cls):
+        def __setitem__(self, key, value):
+            pass
+
+        def __getitem__(self, key, value):
+            raise KeyError(key)
+
+        def __delitem__(self, key):
+            raise KeyError(key)
+
+        def __iter__(self):
+            return iter(())
+
+        def __len__(self):
+            return 0
+
+    d = Func(lambda v: Dumped(), lambda w: None, Dummy())
```

Comment:
Dumped.n increases by 1 every time you create an instance and decreases by 1 every time you garbage-collect it.
Without the change in common.py, n would reach 10 and the assertion on line 69 would fail.

Real life example:

```python
d = Func(pickle.dumps, pickle.loads, File(somedir))
mydata = {str(i): numpy.random.random(10e6) for i in range(100)}
d.update(mydata)
```
mydata occupies 8 GB in RAM (8 * 10e6 * 100).
d.update(mydata) pickles each element and writes it to disk.
Before this PR, you will first create a list of 100 pickles, then write them to disk, and finally release them all at once. You'll observe a buildup of memory while update() is running, going up to 16GB (8GB for mydata + 100\*80MB) and then dropping down all of sudden back to 8GB.
After this PR, update() keeps in memory no more than two pickled arrays at any given time, resulting in a peak memory usage of 8.16 GB (8 GB for mydata + 2\*80MB).

As noted in #63, distributed is not affected by this issue.